pageserver/src/tenant/delete.rs

```diff
@@ -0,0 +1,545 @@
+            Err(anyhow::anyhow!(
+                "failpoint: tenant-delete-before-create-local-mark"
+            ))?
+        });
+
+        create_local_delete_mark(conf, &tenant.tenant_id)
+            .await
+            .context("local delete mark")?;
+
+        fail::fail_point!("tenant-delete-before-background", |_| {
+            Err(anyhow::anyhow!(
+                "failpoint: tenant-delete-before-background"
+            ))?
+        });
+
+        Ok(())
+    }
+
+    fn mark_in_progress(&mut self) -> anyhow::Result<()> {
+        match self {
+            Self::Finished => anyhow::bail!("Bug. Is in finished state"),
+            Self::InProgress { .. } => { /* We're in a retry */ }
+            Self::NotStarted => { /* Fresh start */ }
+        }
+
+        *self = Self::InProgress;
+
+        Ok(())
+    }
+
+    pub async fn should_resume_deletion(
+        conf: &'static PageServerConf,
+        remote_storage: Option<&GenericRemoteStorage>,
+        tenant: &Tenant,
+    ) -> Result<Option<DeletionGuard>, DeleteTenantError> {
+        let acquire = |t: &Tenant| {
+            Some(
+                Arc::clone(&t.delete_progress)
+                    .try_lock_owned()
+                    .expect("we're the only owner during init"),
+            )
+        };
+
+        let tenant_id = tenant.tenant_id;
+        // Check local mark first, if its there there is no need to go to s3 to check whether remote one exists.
+        if conf.tenant_deleted_mark_file_path(&tenant_id).exists() {
+            return Ok(acquire(tenant));
+        }
+
+        // If remote storage is there we rely on it
+        if let Some(remote_storage) = remote_storage {
+            let remote_mark_path = remote_tenant_delete_mark_path(conf, &tenant_id)?;
+
+            let attempt = 1;
+            loop {
+                match remote_storage.download(&remote_mark_path).await {
+                    Ok(_) => return Ok(acquire(tenant)),
+                    Err(e) => {
+                        if matches!(e, DownloadError::NotFound) {
+                            return Ok(None);
+                        }
+                        if attempt > SHOULD_RESUME_DELETION_FETCH_MARK_ATTEMPTS {
+                            return Err(anyhow::anyhow!(e))?;
+                        }
+
+                        warn!(
+                            "failed to fetch tenant deletion mark at {} attempt {}",
+                            &remote_mark_path, attempt
+                        )
+                    }
+                }
+            }
+        }
+
+        Ok(None)
+    }
+
+    pub(crate) async fn resume(
+        guard: DeletionGuard,
+        tenant: &Arc<Tenant>,
+        init_order: Option<&InitializationOrder>,
+        tenants: &'static tokio::sync::RwLock<TenantsMap>,
+        ctx: &RequestContext,
+    ) -> Result<(), DeleteTenantError> {
+        let (_, progress) = completion::channel();
```

Comment:
Could this be explained somehow so that it wouldn't look like it's just ... quick way to call `set_stopping`? (We may have already covered this but) I see following possibilities to enhance this:
- we keep hold of this `_` until the background operation completes
  - if we force it to cover the spawned delete upload ops, we might get completly rid of the shutdown problems (which I still haven't understood how did you solve them)
- you could use `Default::default()` as the argument (or `completion::Barrier::default()`) which would be more explicit that this is not a forgotten (thus requiring less explanation)

Reply:
Ok the other instance has comment above it, just now noticed it.

What was the rationale for not holding up shutdown?

Reply:
Okay, with this rationale it makes sense to me. It should only ever matter to the really near shutdown deleted. And with this rationale we should just have the `allowed_errors` for the deletion tasks in some test config which I'm yet to come across.